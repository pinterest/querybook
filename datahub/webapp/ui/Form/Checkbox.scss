.Checkbox {
    label {
        line-height: 1;
        display: flex;
        flex-direction: row;
        white-space: nowrap;
        position: relative;

        input[type='checkbox'] {
            cursor: pointer;
            appearance: unset;
            border: solid 1px var(--border-color);
            background-color: var(--bg-color);

            margin-right: 8px;
            width: 16px;
            height: 16px;
            outline: none;
        }

        &.checked::before {
            content: ' ';
            background-color: var(--border-color);
            width: 12px;
            height: 12px;
            display: block;
            margin: 2px 0px 0px 2px;
            border-radius: 2px;

            position: absolute;
            left: 1px;
            top: 1px;
        }
    }
}
