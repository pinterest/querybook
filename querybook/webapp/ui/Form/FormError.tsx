import { uniq } from 'lodash';
import React, { useMemo } from 'react';

import { extractFormikError } from 'lib/formik/extract-error';
import { Message } from 'ui/Message/Message';

export const FormError: React.FC<{
    errors: Record<string, any>;
}> = ({ errors }) => {
    const errorStrings = useMemo(
        () => uniq(extractFormikError(errors)),
        [errors]
    );
    if (errorStrings.length === 0) {
        return null;
    }
    return (
        <Message type="error">
            {errorStrings.map((error, idx) => (
                <div key={idx}>
                    <span>{error}</span>
                    <br />
                </div>
            ))}
        </Message>
    );
};
