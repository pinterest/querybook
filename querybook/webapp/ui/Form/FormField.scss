.FormField {
    position: relative;
    display: flex;
    align-items: center;
    margin: 4px;
    padding: 4px;

    &.FormField-stacked {
        flex-wrap: wrap;

        .FormFieldLabelSection {
            margin-bottom: 4px;
        }
        .FormFieldInputSectionRow {
            flex: 1;
            display: flex;
            flex-direction: row;
            .FormFieldInputSection {
                margin-right: 4px;
            }
        }
    }

    .FormFieldLabelSection {
        box-sizing: border-box;
        padding-right: 12px;
        // min-width: 120px;
        cursor: default;
        white-space: nowrap;
    }

    .FormFieldInputSectionRowGroup {
        margin-top: 30px;
    }

    .FormFieldInputSectionRow {
        &:first-of-type {
            margin-top: 0px;
        }
    }

    .FormFieldInputSection {
        flex-grow: 1;
    }

    .FormFieldHelpSection {
        margin-left: 8px;
    }

    .FormFieldErrorSection {
        flex: 1 1 16%;
        padding: 2px 0px 0px 4px;
        box-sizing: border-box;
        color: var(--color-false);
    }

    .FormFieldRequired {
        margin-left: 8px;
    }
}

.FormSectionHeader {
    cursor: default;
    padding-top: 8px;
    span {
        font-size: var(--text-size);
        white-space: nowrap;
        margin-right: 12px;
    }
    hr {
        width: 100%;
        margin: 12px 0px;
    }
}
.popover-content {
    .FormFieldHelpSection {
        p:not(:last-child) {
            margin-bottom: 8px;
        }
        a {
            color: var(--color-accent-text);
            font-weight: var(--bold-font);
        }
    }
}
