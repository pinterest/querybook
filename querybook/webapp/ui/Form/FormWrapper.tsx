import React from 'react';
import styled from 'styled-components';

import { NumFontSize, numFontSizeToString } from 'const/font';

interface ISharedProps {
    minLabelWidth?: string;
    className?: string;
}

const StyledFormWrapper = styled.div.attrs<
    {
        fontSize: string;
        errorFontSize: string;
        sectionFontSize: string;
    } & ISharedProps
>({
    className: ' ',
})`
    font-size: ${(props) => props.fontSize};

    .FormField {
        .FormFieldLabelSection {
            ${(props) =>
                props.minLabelWidth && `min-width: ${props.minLabelWidth}`};
        }

        .FormFieldErrorSection {
            font-size: ${(props) => props.errorFontSize};
        }
    }

    .FormSectionHeader span {
        font-size: ${(props) => props.sectionFontSize};
    }
`;

export const FormWrapper: React.FC<
    {
        size?: NumFontSize;
    } & ISharedProps
> = ({ size = 7, children, ...styleProps }) => {
    const sectionFontSize = numFontSizeToString[Math.max(size - 2, 1)];
    const fontSize = numFontSizeToString[size];
    const errorFontSize = numFontSizeToString[Math.min(size + 1, 10)];

    return (
        <StyledFormWrapper
            fontSize={fontSize}
            sectionFontSize={sectionFontSize}
            errorFontSize={errorFontSize}
            {...styleProps}
        >
            {children}
        </StyledFormWrapper>
    );
};
