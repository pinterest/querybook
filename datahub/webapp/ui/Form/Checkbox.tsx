import React from 'react';
import classNames from 'classnames';
import './Checkbox.scss';

export interface ICheckboxProps {
    className?: string;
    disabled?: boolean;

    title?: string;
    value?: boolean;
    onChange?: (value: boolean) => any;
}

export const Checkbox: React.FunctionComponent<ICheckboxProps> = ({
    className = '',
    disabled,

    title = '',
    value,
    onChange,
    children,
}) => {
    const checkboxClass = classNames({
        [className]: Boolean(className),
        checked: value,
    });

    return (
        <div className="Checkbox">
            <label className={checkboxClass} disabled={disabled}>
                <input
                    type="checkbox"
                    disabled={disabled}
                    checked={value}
                    onChange={(event) => {
                        if (onChange) {
                            onChange(event.target.checked);
                            event.stopPropagation();
                        }
                    }}
                />{' '}
                {title}
                {children}
            </label>
        </div>
    );
};
