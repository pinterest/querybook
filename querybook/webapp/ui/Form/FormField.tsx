import React from 'react';
import clsx from 'clsx';

import { InfoButton } from 'ui/Button/InfoButton';

import './FormField.scss';

export const FormSectionHeader: React.FC = ({ children }) => (
    <div className="FormSectionHeader flex-row">
        <span>{children}</span>
        <hr />
    </div>
);

export type StringOrRender = string | (() => React.ReactNode);

export interface IFormFieldProps {
    stacked?: boolean;
    required?: boolean;
    label?: StringOrRender;
    help?: StringOrRender;
    error?: StringOrRender;
}
export interface IFormFieldSectionProps {
    className?: string;
}

export const FormField: React.FunctionComponent<IFormFieldProps> = ({
    children,
    stacked,
    label,
    help,
    required,
    error,
}) => {
    const labelDOM = label ? (
        <>
            <FormFieldLabelSection>
                {typeof label === 'function' ? label() : label}
            </FormFieldLabelSection>
            {stacked ? <div className="break-flex" /> : null}
        </>
    ) : null;

    const helpDOM = help ? (
        <FormFieldHelpSection>
            {typeof help === 'function' ? help() : help}
        </FormFieldHelpSection>
    ) : null;

    const errorDOM = error ? (
        <FormFieldErrorSection>
            {typeof error === 'function' ? error() : error}
        </FormFieldErrorSection>
    ) : null;

    const requiredIndicator = required ? (
        <i className="FormFieldRequired">*Required</i>
    ) : null;

    // If user uses props to supply label, then auto wrap children to be
    // in input section
    const contentDOM =
        labelDOM || helpDOM || errorDOM ? (
            <FormFieldInputSection>
                {children}
                {errorDOM}
            </FormFieldInputSection>
        ) : (
            children
        );

    return (
        <div
            className={clsx({
                FormField: true,
                'FormField-stacked': stacked,
            })}
        >
            {labelDOM}
            {contentDOM}
            {helpDOM}
            {requiredIndicator}
        </div>
    );
};

const FormFieldLabelSection: React.FunctionComponent<IFormFieldSectionProps> = ({
    children,
    className = '',
}) => <div className={`FormFieldLabelSection ${className}`}>{children}</div>;

export const FormFieldInputSectionRowGroup: React.FunctionComponent<IFormFieldSectionProps> = ({
    children,
    className = '',
}) => (
    <div className={`FormFieldInputSectionRowGroup ${className}`}>
        {children}
    </div>
);

export const FormFieldInputSectionRow: React.FunctionComponent<IFormFieldSectionProps> = ({
    children,
    className = '',
}) => <div className={`FormFieldInputSectionRow ${className}`}>{children}</div>;

export const FormFieldInputSection: React.FunctionComponent<IFormFieldSectionProps> = ({
    children,
    className = '',
}) => <div className={`FormFieldInputSection ${className}`}>{children}</div>;

export const FormFieldHelpSection: React.FunctionComponent<IFormFieldSectionProps> = ({
    children,
    className = '',
}) => (
    <div className={`FormFieldHelpSection flex-center ${className}`}>
        <InfoButton
            layout={['bottom', 'right']}
            popoverClassName="FormFieldHelpSection"
        >
            {children}
        </InfoButton>
    </div>
);

export const FormFieldErrorSection: React.FunctionComponent<IFormFieldSectionProps> = ({
    children,
    className = '',
}) => <div className={`FormFieldErrorSection ${className}`}>{children}</div>;
