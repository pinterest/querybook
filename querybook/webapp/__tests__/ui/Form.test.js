import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';

import {
    FormField,
    FormFieldErrorSection,
    FormFieldHelpSection,
    FormFieldInputSection,
    FormFieldInputSectionRow,
    FormFieldInputSectionRowGroup,
} from '../../ui/Form/FormField';
import { FormWrapper } from '../../ui/Form/FormWrapper';

it('renders without crashing', () => {
    shallow(<FormField />);
});
it('renders without crashing', () => {
    shallow(
        <FormField>
            <FormFieldInputSectionRowGroup>
                <FormFieldInputSectionRow>
                    <FormFieldInputSection>Test</FormFieldInputSection>
                </FormFieldInputSectionRow>
                <FormFieldHelpSection>Test</FormFieldHelpSection>
                <FormFieldErrorSection>Test</FormFieldErrorSection>
            </FormFieldInputSectionRowGroup>
        </FormField>
    );
});
it('renders without crashing', () => {
    shallow(<FormWrapper />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<FormField />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
    it('matches snapshot', () => {
        const wrapper = shallow(<FormWrapper />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});
