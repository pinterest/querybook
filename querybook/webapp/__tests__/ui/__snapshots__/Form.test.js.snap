// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<div
  className="FormField"
/>
`;

exports[`matches enzyme snapshots matches snapshot 2`] = `
<FormWrapper__StyledFormWrapper
  errorFontSize="var(--small-text-size)"
  fontSize="var(--text-size)"
  sectionFontSize="var(--med-text-size)"
/>
`;
